Make Python init more language native

daytonb opened the pull request:
This pull request is an effort to implement a more language-native way to get the "module" function in Python scripts. It is the Python portion of #483. I first proposed this pull request in the mailing list. Below is a description of the plan from that conversation.

For clarity I'll use "Python-module" to refer to a Python script defining functions/classes  that you `import` in Python code and "environment-module" to refer to a TCL environment module that you `module load`/`module unload`/etc.

The current documentation for Python initialization (https://modules.readthedocs.io/en/latest/module.html#examples-of-initialization) has 

```
import os
exec(open("/usr/share/Modules/init/python.py").read(), globals())
module("load", "modulefile", "modulefile", "...")
```

A more Pythonic approach would be to `import` this `module` function somehow. Currently, you _can_ do that with as follows

```
import sys
sys.path.append("/usr/share/Modules/init")
from python import module
module("load", "modulefile", "moduelfile", "...")
```

The issues with this are that you have to modify the `sys.path` inside the script and since the Python-module `import`-ed is called "python". The name makes it look like `module` is a function from some core Python-module from the Python project.

In this pull request I will
1. rename the Python initialization file to `env_module.py` (this is what lmod named theirs by the way) and
2. to have the shell initialization in `tcl/subcmd.tcl.in` add the folder of initilization scripts to the PYTHONPATH environment variable.

These changes would change Python initialization to

```
from env_module import module
module("load", "modulefile", "moduelfile", "...")
```

which cleans up the way the Python script imports the `module` function and communicates more clearly that it comes from a 3rd-party Python-module related to environment modules.

Commit:
Make it more Pythonic to import env_modules.module

Renaming python.py to env_modules.py. On Linux, symlink python.py to
point to env_modules.py for backward compatibility. Windows doesn't
typically support unix-style symlinks, but we don't need to worry about
that since the Windows installations of Environment Modules never had
python.py.

Signed-off-by: Byron Boulton <bboulton@ara.com>

Commit:
doc: describe initialization of PYTHONPATH

Signed-off-by: Byron Boulton <bboulton@ara.com>
Signed-off-by: Xavier Delaruelle <xavier.delaruelle@cea.fr>

Commit:
Implement PYTHONPATH initialization

Adding the initdir to PYTHONPATH allows users to "import" the "module"
function in Python scripts.

Signed-off-by: Byron Boulton <bboulton@ara.com>

daytonb commented:
At this time there are a couple of issues I know I need to address:
1. Add tests - I'm guessing we probably want some tests added
2. Resolve test failures - I ran the quicktests locally and got 42 failures, but I couldn't see how you see which tests failed in order to figure out what I broke
3. Have someone do a code a documentation review
4. Change the `python.py` file to an install-time symlink on Linux:
    * In the mailing list there was discussion that for compatibility the installation process on Linux should symlink the new `env_modules.py` to `python.py`, and that no such compatibility in neccessary for Windows because the Windows installs have never had the `python.py` file.

xdelaruelle commented on configure:
```diff
@@ -237,6 +239,8 @@ Optional Features:
   --enable-set-manpath    set mandir to MANPATH in init scripts [yes]
   --enable-append-manpath append rather prepend mandir to MANPATH [no]
+  --enable-set-pythonpath set initdir to PYTHONPATH in init scripts [yes]
+  --enable-append-manpath append rather prepend initdir to PYTHONPATH [no]
```
small typo here: should mention --enable-append-pythonpath

xdelaruelle commented on init/env_modules.py.in:
```diff
@@ -0,0 +1,33 @@
+import os, sys, re, subprocess
```
I would prefer to have the `python.py.in` moved in the repository as `env_modules.py.in` to avoid having 2 different files. The symlink will provide compatibility at installation time.

daytonb replied:
I saw in the mailing list conversation that you noted the Windows artifacts haven't included python.py, so you're right that a symlink on the Linux installations makes more sense.

xdelaruelle commented on tcl/subcmd.tcl.in:
```diff
@@ -2157,6 +2157,13 @@ proc cmdModuleAutoinit {} {
    }
 
+   # add Modules init directory to PYTHONPATH if enabled
+   @setpythonpath@if {![isEnvVarDefined PYTHONPATH]} {
```
that branch of the `if` is not needed (only add path needed, no setenv as add path will do that if variable does not exist yet)

daytonb replied:
Ah. I see. That seems pretty clear now that I think about that more. The similar chunks above which modify PATH and MANPATH aren't perfect examples for modifying PYTHONPATH. I'm not sure I follow the purpose of the `--ignore-refcount` part of modifying PATH. Thank goodnesss PYTHONPATH doesn't have to handle ensuring we have an "empty" `:` like MANPATH.

xdelaruelle reviewed (changes requested):
Many thanks for this pull request. This is already in a very good shape. See review comments for the things to adapt.

Regarding tests:

* the `script/mt` utility will guide you to locate where the test failures come from
* a `grep -R setbinpath testsuite/` run at the root of the repository will help you to see where new tests should be added (you may largely inspire from these setbinpath tests)

daytonb commented:
I've addressed the issues noted above except the failing tests. I still need to:
* find and fix the failing tests
* add new tests for the pythonpath configure options

xdelaruelle reviewed (changes requested):
Please see my comments for the few changes to apply.

daytonb commented:
I pushed up some fixes. I want to rebase/squash them, but I need to sign off for now.

xdelaruelle committed:
doc: extend --enable-set-pythonpath documentation

Signed-off-by: Xavier Delaruelle <xavier.delaruelle@cea.fr>

xdelaruelle committed:
ts: adapt tests for --enable-set-pythonpath

Signed-off-by: Xavier Delaruelle <xavier.delaruelle@cea.fr>

xdelaruelle reviewed (approved):
Many thanks @daytonb for this contribution.

I have extended the documentation and adapted the tests.

I will be able to merge this pull request after CI run.
